libnetwork/portmapper/mapper.go

```diff
@@ -1,3 +1,5 @@
+//go:build windows
```

Comment:
I know that on Linux only the bridge driver uses this package, but honestly I'd prefer to keep both packages clearly separated to make sure we can easily test and evolve the portmapper. For instance, this would be helpful to add nftables support.

Also, looking at it, there are still some bits related to Linux like `StartProxy`, contradicting what the commit message says.

Generally speaking, it's better to not move code and evolve it at the same time. Otherwise it makes your life and reviewers' life harder (eg. when you have to rebase, etc...).

Reply:
> I know that on Linux only the bridge driver uses this package, but honestly I'd prefer to keep both packages clearly separated to make sure we can easily test and evolve the portmapper. For instance, this would be helpful to add nftables support.

On Linux yes, only the bridge driver. But there's also Windows - which is almost completely different, but shared the same code. `bridge` had half of the bridge-specific iptables logic, with per-port logic living in `iptables`. We've discussed moving driver-specific logic into the driver several times, it's the right place for it.

The bridge's requirements have changed with the option to have no-NAT, iptables hasn't changed, but the implementation of forwarding/DNAT/masquerading etc was in `iptables.Forward` - behind two layers of indirection in the Linux & Windows `PortMapper` that would also both need to change.

`PortMapper` has a layer of indirection so that Windows can stub out most of the logic. If I'd made the changes in PortMapper, I'd have been changing the Windows code (which is hard to test and doesn't need to change yet, because it doesn't have IPv6 - when it does, the changes needed will inevitably be quite different to what was needed for Linux).

The `PortMapper` was also designed to deal with IPv4/IPv6 separately, which doesn't work.

So, `PortMapper` had to change. The best evolution for it is elimination - with `bridge` composing tools from `portallocator`/`iptables` and `portmapper` docker-proxy code directly.

In a first commit I called `iptables.Forward` directly from `bridge` code instead of indirectly through the `PortMapper`. In a second change I moved its bridge-specific logic into the bridge package - which meant using `bridge.iptRule` to be consistent with the rest of the bridge code. (`bridge.iptRule` itself isn't ideal but it can be improved separately, for all of the `bridge` code that uses it.)

> Also, looking at it, there are still some bits related to Linux like `StartProxy`, contradicting what the commit message says.

Right, the comment says "So, remove linux-specific portmapper code". It should be "So, remove linux-specific PortMapper code".

> Generally speaking, it's better to not move code and evolve it at the same time. Otherwise it makes your life and reviewers' life harder (eg. when you have to rebase, etc...).

Yes, agreed. But I think the alternative to what I've described above is not changing it, or making the messy Windows & Linux implementation more messy, so that it can still be used by both drivers to do quite different things. Is there anything specific I can do to improve things?